### PR TITLE
allow users to set sample rate and add per iteration latency breakdowns

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53148 clang format reducer and logger files
* #53162 fix data type logging
* #53149 add is_multi_device_module logging field
* **#53145 allow users to set sample rate and add per iteration latency breakdowns**

add a new API to allow users to set sample rate for runtime stats, also add per iteration latency breakdowns to DDPLoggingData struct. e.g.
if users set sample rate to be 1, they can analyze per iteration latency change over time (not avged)

Differential Revision: [D26763957](https://our.internmc.facebook.com/intern/diff/D26763957/)

### CODE REVIEW

File: torch/lib/c10d/reducer.hpp [hunk 1]
```diff
@@ -379,6 +385,9 @@ class Reducer {
   void record_backward_comm_start_time();
   void record_backward_comm_end_time();
 
+  int get_ddp_runtime_logging_sample_rate();
+  int ddp_runtime_logging_sample_rate_ = 0;
```
Comment (SciPioneer) [inline torch/lib/c10d/reducer.hpp:389]: Why not directly use kDDPRuntimeLoggingSampleRate?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -1277,3 +1277,27 @@ def _set_params_and_buffers_to_ignore_for_model(
...
+        This interface allows users to set sample_rate of collecting
+        runtime stats. The runtime stats will be recorded for the
+        first 10 iterations, after 10 iteratons runtime stats will be
+        recorded once every "sample_rate" training iterations. In
+        default, runtime stats are recorded for the first 10 iterations,
+        after 10 iterations runtime stats are recorded once every
+        "kDDPRuntimeLoggingSampleRate=100" training iterations.
+        """
+        if sample_rate <= 0:
+            raise ValueError("DDP runtime logging sample rate should be larger than 0")
```
Comment (SciPioneer) [inline torch/nn/parallel/distributed.py:1302]: nit: larger/greater

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -1277,3 +1277,27 @@ def _set_params_and_buffers_to_ignore_for_model(
...
+    def get_ddp_logging_data(self):
+        r"""
+        This interface can be called after DistributedDataParallel() is
+        constructed. It returns DDPLoggingData for debugging and analysis.
+        More detailed explanation of the fields in DDPLoggingData are in
+        ``torch/c10/util/Logging.h``.
+        """
+        return self.logger._get_ddp_logging_data()
+
+    def set_ddp_runtime_logging_sample_rate(self, sample_rate):
```
Comment (SciPioneer) [inline torch/nn/parallel/distributed.py:1291]: The term "sample rate" actually can be a bit misleading. Normally sampling rate is a fractional number between 0 and 1. I think "sampling period" or "sampling interval" can  be a better term here.

**Review by wayi1 — APPROVED** [review]

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -1305,3 +1305,27 @@ def _set_params_and_buffers_to_ignore_for_model(
...
+        r"""
+        This interface allows users to set sample_rate of collecting
+        runtime stats. The runtime stats will be recorded for the
+        first 10 iterations, after 10 iteratons runtime stats will be
+        recorded once every "sample_rate" training iterations. In
+        default, runtime stats are recorded for the first 10 iterations,
+        after 10 iterations runtime stats are recorded once every
+        "kDDPRuntimeLoggingSampleRate=100" training iterations.
+        """
+        if sample_rate <= 0:
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:1329]: Should we check that the arg is an int and greater/equal to 1? Something like 0.5 for example doesn't make sense (and wouldn't that be cast down to 0 if it's passed into cpp as an int?)

**Pull request closed without merging**